### PR TITLE
Add several lints into `unused` lint group

### PR DESCRIPTION
**Pull request by petrochenkov**
Also a couple of obsolete (not reported) lints are removed.

r? @oli-obk

### CODE REVIEW

File: src/librustc_passes/consts.rs [hunk 1]
```diff
@@ -140,8 +140,7 @@ impl<'a, 'gcx> CheckCrateVisitor<'a, 'gcx> {
                     self.tcx.lint_node(CONST_ERR,
                                        expr.id,
                                        expr.span,
-                                       &format!("constant evaluation error: {}. This will \
-                                                 become a HARD ERROR in the future",
+                                       &format!("constant evaluation error: {}",
```
Comment (petrochenkov) [inline src/librustc_passes/consts.rs]: @oli-obk 
I initially wanted to move `const_err` into the `future_incompatible` group based on this comment, but it looks like this comment was actually outdated and this lint is not going to be an error?

**Comment by petrochenkov** [comment]
I've also audited the remaining ungrouped lints.
In principle new lint groups can be created for them, but it looks like fine-grained lint grouping isn't used even in clippy. cc clippy people just in case @mcarton @llogiq

**Unused++.**
These can also be quite reasonably added into the `unused` group, but less obviously than those I added into it in this PR.
```
STABLE_FEATURES
RENAMED_AND_REMOVED_LINTS
UNKNOWN_LINTS
UNUSED_COMPARISONS
```

**Bad style++.**
Probably can be added into the `bad_style` group, but it currently consists only of casing-related lints.
```
NON_SHORTHAND_FIELD_PATTERNS
WHILE_TRUE
```

**Future compatibility++.**
Errors that are reported as lints for some reasons unknown to me.
See the question in https://github.com/rust-lang/rust/pull/45424#discussion_r146084996 as well.
```
CONST_ERR // ?
UNKNOWN_CRATE_TYPES // Deny-by-default
NO_MANGLE_CONST_ITEMS // Deny-by-default
NO_MANGLE_GENERIC_ITEMS
```

**Restrictions.**
Something generally reasonable that can be prohibited if necessary.
```
BOX_POINTERS // Allow-by-default
UNSAFE_CODE // Allow-by-default
UNSTABLE_FEATURES // Allow-by-default
MISSING_DOCS // Allow-by-default
MISSING_COPY_IMPLEMENTATIONS // Allow-by-default
MISSING_DEBUG_IMPLEMENTATIONS // Allow-by-default
```

**Pedantic.**
Something not bad enough to always report/fix.
```
UNUSED_RESULTS // Allow-by-default
UNUSED_IMPORT_BRACES // Allow-by-default
UNUSED_QUALIFICATIONS // Allow-by-default
TRIVIAL_CASTS // Allow-by-default
TRIVIAL_NUMERIC_CASTS // Allow-by-default
VARIANT_SIZE_DIFFERENCES // Allow-by-default
UNIONS_WITH_DROP_FIELDS
```

**Obvious mistakes.**
Prevent foot shooting, some can become hard errors in principle.
```
OVERFLOWING_LITERALS
EXCEEDING_BITSHIFTS // Deny-by-default
UNCONDITIONAL_RECURSION
MUTABLE_TRANSMUTES // Deny-by-default
IMPROPER_CTYPES
PLUGIN_AS_LIBRARY
PRIVATE_NO_MANGLE_FNS
PRIVATE_NO_MANGLE_STATICS
```

**General purpose lints.**
```
WARNINGS
DEPRECATED
```

File: src/librustc_passes/consts.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_passes/consts.rs]: Yea. These are essentially bug

File: src/librustc_lint/builtin.rs [hunk 2]
```diff
@@ -621,12 +621,6 @@ impl EarlyLintPass for AnonymousParameters {
     }
 }
 
-declare_lint! {
-    DEPRECATED_ATTR,
```
Comment (oli-obk) [inline src/librustc_lint/builtin.rs]: Just removing these lints will cause users who mention them in allow/warn/deny statements to produce warnings about unknown lints. Does rustc have a removed lint list like clippy does?

**Review by oli-obk — CHANGES_REQUESTED** [review]
The lint store has a function register_removed

**Comment by oli-obk** [comment]
I would like fine grained grouping. I think we have an open issue in clippy. People often complain that they are overwhelmed by the style lints, when they want to get to the bug lints first

File: src/librustc_lint/builtin.rs (hunk 2 above)
Comment (petrochenkov) [reply src/librustc_lint/builtin.rs]: Yes, there's a list of removed lints, I'll add these these into it.

**Comment by petrochenkov** [comment]
Rebased and updated with a fix for https://github.com/rust-lang/rust/pull/45424#discussion_r146107544

**Comment by oli-obk** [comment]
Lgtm now. Do we have an issue to discuss the fine grained lint groups and not clear cut "unused"-lints? https://github.com/rust-lang/rust/pull/45424#issuecomment-338345084 as an issue would be good I think

@bors r+

**Comment by petrochenkov** [comment]
@oli-obk 
>Do we have an issue to discuss the fine grained lint groups and not clear cut "unused"-lints? 

I've created https://github.com/rust-lang/rust/issues/45615 for this.

**Comment by kennytm** [comment]
@bors retry #43283 (`arm-android` test_process_output_fail_to_start 30min no output)

```
[01:10:10] test process::tests::test_process_output_fail_to_start has been running for over 60 seconds


No output has been received in the last 30m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received

The build has been terminated
```

**Comment by kennytm** [comment]
The use of `:vis` have broken clippy. But should the feature gate error be emitted when compiling clippy, as a user of the macro instead of the provider? cc #41022 @durka.

<details>

```
[00:34:29]    Compiling clippy_lints v0.0.166 (file:///checkout/src/tools/clippy/clippy_lints)
[00:34:29] error: :vis fragment specifier is experimental and subject to change (see issue #41022)
[00:34:29]  --> <declare_lint macros>:1:3
[00:34:29]   |
[00:34:29] 1 | ( $ vis : vis $ NAME : ident , $ Level : ident , $ desc : expr ) => (
[00:34:29]   |   ^^^^^^^^^^^
[00:34:29]   |
[00:34:29]   = help: add #![feature(macro_vis_matcher)] to the crate attributes to enable
[00:34:29] 
[00:34:29] error: :vis fragment specifier is experimental and subject to change (see issue #41022)
[00:34:29]  --> <declare_lint macros>:1:3
[00:34:29]   |
[00:34:29] 1 | ( $ vis : vis $ NAME : ident , $ Level : ident , $ desc : expr ) => (
[00:34:29]   |   ^^^^^^^^^^^
[00:34:29]   |
[00:34:29]   = help: add #![feature(macro_vis_matcher)] to the crate attributes to enable
[00:34:29] 
[00:34:30] error: aborting due to previous error
[00:34:30] 
[00:34:30] error: Could not compile `clippy_lints`.
[00:34:30] warning: build failed, waiting for other jobs to finish...
[00:34:30] error: aborting due to previous error
[00:34:30] 
[00:34:30] error: Could not compile `clippy_lints`.
[00:34:30] warning: build failed, waiting for other jobs to finish...
[00:34:41] error: build failed
```

</details>

**Comment by durka** [comment]
Hmm. Ideally the writer of the macro should need the feature gate, not the user. But AFAIK pretty much all checking of macros is left until the last possible second. For example the current version of `error_chain` has an [undefined capture](https://github.com/rust-lang-nursery/error-chain/blob/48c18a9747bb31b663d990f8971963706a4652d7/src/impl_error_chain_kind.rs#L73) and nobody noticed until [yesterday](https://botbot.me/mozilla/rust/2017-10-29/?msg=92862308&page=3).

**Comment by durka** [comment]
Let's just stabilize `:vis` and remove the issue 😛

**Comment by Manishearth** [comment]
Uh, please at least try to fix tools that break (even if you mark it as broken in the interim).

In this case for some reason `#![feature(macro_vis_matcher)]` does not fix the problem in clippy; it continues to give the same error despite being allowed already.

**Comment by Manishearth** [comment]
Oh, nvm, I made a mistake. Sorry.

**Pull request merged**